File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:    
    int maxSubarrayLength(vector<int>& nums, int k) {
        
        int n = nums.size(), maxF = 0;
        
        int i = 0, j = 0, res = 0;
        
        unordered_map<int,int>freq;
        
        while(j < n){
            
            ++freq[nums[j]];
            
            maxF = max(maxF, freq[nums[j]]);
            
            while(i < j && maxF > k){
                
                if(maxF == freq[nums[i]])   --maxF;
                
                --freq[nums[i]];
                
                ++i;
                
            }
            
            res = max(res, j - i + 1);
            
            ++j;
            
        }
        
        return res;
        
    }
};

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        Map<Integer, Integer>map = new HashMap<>();
        int i = 0, j = 0, maxF = 0, n = nums.length;
        int res = 0;
        
        while(j < n){
            map.put(nums[j], map.getOrDefault(nums[j], 0) + 1);
            
            while(i < j && map.get(nums[j]) > k){
                map.put(nums[i], map.get(nums[i]) - 1);
                ++i;
            }
            
            res = Math.max(res, j - i + 1);
            ++j;
        }
        
        return res;
    }
}